tripal/tripal
# Support Drupal 11.0.x + 11.1.x

# New Feature

### Closes #1898 

### Tripal Version: 4.x

## Description
Changes to support Drupal 11.0.x-dev and 11.1.x-dev.

~~The docker has been updated to not attempt to install tripal_layout on Drupal 11 since we are waiting on https://www.drupal.org/project/field_group_table/issues/3430507. We are no longer trying to get Tripal Layout D11 compatible in this PR as that issue has been RTBC for months 😒 Instead we are focusing on the other Tripal Core modules so we don't get further behind 🙈~~

Field Group Table has finally been made D11 compliant! This PR thus intends to bring all Tripal modules to be compatible with D11 :-)

Status: All automated tests are passing, warnings have been dealt with and deprecation notices have been addressed. However, a thorough testing of a D11 site through the UI has not yet been done.
Basic testing for D11 was done in comment https://github.com/tripal/tripal/pull/1953#issuecomment-2479180184

## Testing?
1. Build a docker on this branch, e.g. 
```
docker build \
  --tag=tripaldocker:testing-1953 \
  --build-arg drupalversion="11.0.x-dev" \
  --build-arg postgresqlversion="16" \
  --build-arg chadoschema="chado" \
  --progress=plain \
  ./
git='where you have this checked out'
docker run \
  --publish=8080:80 -tid \
  --name=1898 \
  --volume=$git:/var/www/drupal9/web/modules/contrib/tripal \
  tripaldocker:testing-1953
```
2. Go to your site and start doing stuff and find out what is broken,
    e.g. it looks like autocomplete is not working for infraspecific type.

3. Make sure the javascript fix from commit https://github.com/tripal/tripal/pull/1953/commits/3ad8feb7897052fb5672f3bdf47331f6e54e5555 
  works on both Drupal 11.0 and 10.x, you just need to go to the GFF3 importer
  and upload a file, you don't need to actually run the importer.

## Tasks
- [x] **Fix autocomplete**
`Drupal\Core\Database\DatabaseExceptionWrapper: SQLSTATE[22025]: Invalid escape sequence: 7 ERROR: invalid escape string HINT: Escape string must be empty or one character.: SELECT "CVT"."name" AS "name", "CVT"."definition" AS "definition", "CVT"."cvterm_id" AS "cvterm_id", "DBX"."accession" AS "accession", "CV"."name" AS "CV_name" FROM chado.cvterm "CVT" INNER JOIN chado.cv "CV" ON "CV".cv_id = "CVT".cv_id INNER JOIN chado.dbxref "DBX" ON "DBX".dbxref_id = "CVT".dbxref_id INNER JOIN chado.db "DB" ON "DB".db_id = "DBX".db_id WHERE ("DB"."name" = :db_condition_placeholder_0) AND ("CVT"."name"::text LIKE :db_condition_placeholder_1 ESCAPE '\\'); Array ( [:db_condition_placeholder_0] => CO_010 [:db_condition_placeholder_1] => subspecies% ) in Drupal\tripal_chado\Plugin\TripalIdSpace\ChadoIdSpace->getTerms() (line 426 of /var/www/drupal/web/modules/contrib/tripal/tripal_chado/src/Plugin/TripalIdSpace/ChadoIdSpace.php).`
Reference for postfix `\\\\` https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Database%21Query%21Condition.php/property/Condition%3A%3AconditionOperatorMap/8.9.x
Reference for escapeLike() https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Database%21Connection.php/function/Connection%3A%3AescapeLike/8.9.x

- [x] "Upload file" button on importer not working
  An error inside `tripal/js/TripalUploader.js` 
  This line `var func_name = ($.isFunction($.fn.live)) ? 'live' : 'on';`
  causes this error
![2024-08-20_js-1](https://github.com/user-attachments/assets/6b3eaee8-5eae-459b-952c-74913d23af8c)
See commit https://github.com/tripal/tripal/pull/1953/commits/3ad8feb7897052fb5672f3bdf47331f6e54e5555 for a fix

- [x] Fix test ChadoVariablesAPITest::testChadoGenerateVariables

```
Drupal\Tests\tripal_chado\Functional\api\ChadoVariablesAPITest
  ❌ testChadoGenerateVariables 72923 ms
     ┐ 
     ├ The residues should be missing by default.
     ├ Failed asserting that true is false.
     │ 
     │ /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/api/ChadoVariablesAPITest.php:119
     ┴ 
```

- [x] Fix test GFF3ImporterTest::testGFFImporterSimpleTest

```
Drupal\Tests\tripal_chado\Functional\GFF3ImporterTest
  ❌ testGFFImporterSimpleTest 2910 ms
     ┐ 
     ├ This test printed output: 
     ├ Warning: Undefined array key "is_target" in /var/www/drupal/web/modules/contrib/tripal/tripal_chado/src/Plugin/TripalImporter/GFF3Importer.php on line 3216
     ├ 
     ├ Warning: Placeholders must begin with one of the following "@", ":" or "%", invalid placeholder (!landmark) with string: "Landmark/seqid !landmark contains invalid
     ├         characters. Only characters included in this regular expression are
     ├         allowed [a-zA-Z0-9.:^*$@!+_?-|]" in /var/www/drupal/web/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 217
     │ 
     ┴ 
```
